elastic/apm-aws-lambda
# End-to-end testing using SAM CLI and locally run Lambdas

#### Motivation / Summary
The work leading to this PR was initiated based on PR #67, which is effectively replaced by the following Pull Request. The objective is to provide contributors with an automated method to test the APM data flow end-to-end, in the context of AWS lambda functions. 

#### Files
This PR adds a subfolder `e2e-testing` to the extension folder. This subfolder contains a Go script `e2e_test.go`, which is the end-to-end test method itself. Alongside the Go script, an additional `.gitignore` three folders containing the lambda function to be built are provided. These three folders correspond to the 3 languages for which the APM agent currently supports the AWS Lambda extension : Javascript (Node), Python and Java.

#### Checklist

- [x] Make the execution of the Lambda functions fully local
- [x] Automate the retrieval and the permission changes on the APM Java agent layer
- [x] Test APM-instrumented application -> Elasticsearch data flow through a UUID-named transaction
- [x] Use the Golang testing framework instead of the `main` package
- [x] By default, remove the E2E test from the default testing suite (an environment variable has to be set explicitely in order to run the test
- [ ] Add an additional check to confirm that the APM extension is running and healthy in the test containers 
- [ ] Parallelise the tests for each of the 3 languages
- [ ] Make the test usable with both a local Elasticsearch instance and a Cloud instance
- [x] Close PR #67 

#### How to test

- Download / Clone the repository
- In `apm-lambda-extension/e2e-testing`, create a .env file and set the following environment variables : 
```
RUN_E2E_TESTS=true
ELASTIC_CLOUD_ID=<The Cloud ID of your Elastic Cloud tenant>
ELASTIC_CLOUD_API_KEY=<The API Key of your Elastic Cloud tenant>
ELASTIC_APM_SERVER_URL=<The URL of the APM server linked to your Elastic Cloud tenant>
ELASTIC_APM_SERVER_TOKEN=<The Token of the APM server linked to your Elastic Cloud tenant>
FORCE_REBUILD_LOCAL_SAM_LAMBDAS=false
```
Note that without the `.env` file, or if `RUN_E2E_TESTS` is missing, the test will be skipped.
- Run the test
```
cd e2e-testing
go test
```
